### DISCUSSION TITLE
Why buffers are not synchronized in zero3?

### DISCUSSION THREAD

**Discussion by kiradiso** [deepspeedai/DeepSpeed#6829] [Q&A]
Hellow, i launch my model by `accelerate launch` with zero3 and found all parameters are partitioned. For my context, i apply `processor.vision_tokenizer` only for inference which is conducted by following code:
```
with deepspeed.zero.GatheredParameters(processor.vision_tokenizer.parameters(), modifier_rank=None):
    inputs = processor.vision_tokenizer(pixel_values)
```
However, i found `inputs` varies from different ranks (with same `pixel_values`). After some checks, i found the buffers of `nn.BatchNorm3d` are different (runner_vars/means). Only the buffers of rank 0 is correct.
So my questions are:
1. how to make these buffers consistent across ranks?
2. Whether exists an more better way to conduct inference for `processor.vision_tokenizer`, which is invariant to the transformer for training.

**Comment by tjruwase** [comment]
@kiradiso, thanks for your questions
1. Try setting `modifier_rank=0` to get consistent buffers. See [here](https://www.deepspeed.ai/tutorials/zero/#allocating-massive-megatron-lm-models). 
2. Please clarify what you mean by better? Do you mean low latency, higher throughput, memory efficient, etc.

**Reply by kiradiso** [reply]
Thanks for your answer
1. I find that setting `modifier_rank=0` is not worked for me. My code is like this:
Program is launched by `accelerate launch --config_file={a zero3 config} ...`
In main program, i have：
```
accelerator = Accelerator(
        gradient_accumulation_steps=1,
        mixed_precision="bf16",
        log_with="wandb",
        project_dir={a logging_dir},
        split_batches=True,
    )
image_tokenizer = AutoModel.from_pretrained("BAAI/Emu3-VisionTokenizer", trust_remote_code=True)
with deepspeed.zero.GatheredParameters(image_tokenizer.parameters(), modifier_rank=0):
    print(image_tokenizer.encoder.time_res_stack[0].norm1._buffers, torch.distributed.get_rank())
```
The outputs are like this:
```
OrderedDict([('running_mean', tensor([ 8.1875,  6.7812,  5.2812, 11.0625], device='cuda:0', dtype=torch.bfloat16)), 
('running_var', tensor([4.5312, 1.6484, 2.1406, 5.8438], device='cuda:0', dtype=torch.bfloat16)), 
('num_batches_tracked', tensor(650000, device='cuda:0'))]) 0

OrderedDict([('running_mean', tensor([0., 0., 0., 0.], device='cuda:1', dtype=torch.bfloat16)), 
('running_var', tensor([1., 1., 1., 1.], device='cuda:1', dtype=torch.bfloat16)), 
('num_batches_tracked', tensor(0, device='cuda:1'))]) 1
```
One solution is moving the definition of `image_tokenizer` before the definition of `accelerator`, then, `image_tokenizer` is not partitioned.

2. I don't want `image_tokenizer` to be partitioned, but with my usage, any model defined after the definition of `accelerator ` is partitioned. If i don't use accelerator, i think i can use `deepspeed.zero.Init` as a context manager to actively restrict the scope of auto-partition (i have not tried).

**Reply by tjruwase** [reply]
I think the problem is caused by underlying [HfDeepSpeedConfig](https://huggingface.co/docs/transformers/v4.15.0/en/main_classes/deepspeed#nontrainer-deepspeed-integration) used for managing the DeepSpeed integration. When used in a global context, this imposes the first ds_config on all DeepSpeed interactions. One workaround is to move each model or tokenizer initialization into separate functions to restrict the HfDeepSpeedConfig to function scope instead of global. I realize that Accelerate does not expose HfDeepSpeedConfig but I think the backend operates similarly. 

If you are unable to share your full code with the `Accelerate` call context, then perhaps the following transformer example could explain more: https://gist.github.com/tjruwase/f1adccf087b8fa269ffce2ab91c4f1c6

**Comment by Mikyx-1** [comment]
This is expected behavior, not a bug: ZeRO-3 (and `GatheredParameters`) only ever touches **parameters**, never buffers.

- `zero.Init`'s `_post_init_method` in `partition_parameters.py` only iterates `module.named_parameters(recurse=False)` — there's no `named_buffers()` call anywhere in that file.
- `GatheredParameters` filters its input down to objects with a `ds_id` attribute, which only ZeRO-managed parameters have; buffers never get one, so they're silently skipped.
- Parameter values *are* broadcast from rank 0 at `zero.Init` time (`dist.broadcast(param.data, 0, partition_group)`), but there's no equivalent broadcast for buffers.

So your `BatchNorm3d` running stats are just whatever each rank computed/initialized locally — DeepSpeed has no DDP-style `broadcast_buffers=True` behavior.

I put together a minimal repro to confirm this and verify two fixes (script below, run on 2 GPUs):

```
STEP 1 - Reproduce the bug
[BUG REPRO] max|out_rank0 - out_rankN| per rank: [0.0, 4.065269470214844]

STEP 2 - Fix: broadcast buffers after construction/load
[FIX 1 - broadcast buffers] max|out_rank0 - out_rankN| per rank: [0.0, 0.0]

STEP 3 - Fix: keep the frozen inference-only module outside zero.Init
[FIX 2 - kept outside zero.Init] max|out_rank0 - out_rankN| per rank: [0.0, 0.0]

STEP 4 - Bonus: per-forward-call overhead
[TIMING] avg per-call: GatheredParameters=0.577ms  plain-replica=0.022ms  (26.6x slower with GatheredParameters)
```

**Fix 1 — broadcast buffers yourself**, since DeepSpeed won't:
```python
import deepspeed.comm as dist
for buf in model.buffers():
    dist.broadcast(buf, src=0)
```

**Fix 2 (probably the better one for your case)** — since `vision_tokenizer` is frozen and only used for inference, it never needs ZeRO-3 sharding at all. Build it *outside* the `zero.Init()` context so it stays a full replica on every rank from the start — no buffer-sync bug is possible, and you also skip the per-forward `GatheredParameters` all-gather/release, which the repro shows costs ~27x more per call than a plain replicated forward.

Repro script, in case it's useful to reproduce on your end or extend:

```python
# SPDX-License-Identifier: Apache-2.0
# DeepSpeed Team
import os
import torch
import torch.multiprocessing as mp
import torch.nn as nn

import deepspeed
import deepspeed.comm as dist
from deepspeed.accelerator import get_accelerator

WORLD_SIZE = 2
ZERO3_CONFIG = {"train_micro_batch_size_per_gpu": 1, "zero_optimization": {"stage": 3}}


class VisionTokenizer(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = nn.Conv3d(3, 8, kernel_size=3, padding=1)
        self.bn = nn.BatchNorm3d(8)

    def forward(self, x):
        return self.bn(self.conv(x))


def init_dist(rank, world_size, master_port):
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = str(master_port)
    os.environ["LOCAL_RANK"] = str(rank)
    os.environ["RANK"] = str(rank)
    os.environ["LOCAL_SIZE"] = str(world_size)
    os.environ["WORLD_SIZE"] = str(world_size)
    get_accelerator().set_device(rank)
    deepspeed.init_distributed(dist_backend=get_accelerator().communication_backend_name(),
                                init_method="env://", rank=rank, world_size=world_size)


def build_model_under_zero_init():
    with deepspeed.zero.Init(config_dict_or_path=ZERO3_CONFIG):
        model = VisionTokenizer()
    return model.to(get_accelerator().current_device_name())


def simulate_bad_checkpoint_load(model, rank):
    # mimics only rank 0's loader actually copying pretrained buffer values in
    if rank == 0:
        with torch.no_grad():
            model.bn.running_mean.fill_(5.0)
            model.bn.running_var.fill_(2.0)


def get_shared_input():
    x = torch.randn(1, 3, 4, 8, 8, device=get_accelerator().current_device_name())
    dist.broadcast(x, 0)
    return x


def run_forward_with_gather(model, x):
    model.eval()
    with deepspeed.zero.GatheredParameters(model.parameters(), modifier_rank=None):
        with torch.no_grad():
            return model(x)


def report_diffs(tag, rank, world_size, out):
    gathered = [torch.zeros_like(out) for _ in range(world_size)]
    dist.all_gather(gathered, out.contiguous())
    if rank == 0:
        diffs = [(gathered[0] - gathered[i]).abs().max().item() for i in range(world_size)]
        print(f"{tag} max|out_rank0 - out_rankN| per rank: {diffs}")
    dist.barrier()


def case_bug(rank, world_size, _port):
    init_dist(rank, world_size, 29511)
    model = build_model_under_zero_init()
    simulate_bad_checkpoint_load(model, rank)
    out = run_forward_with_gather(model, get_shared_input())
    report_diffs("[BUG REPRO]", rank, world_size, out)


def case_fix1_broadcast_buffers(rank, world_size, _port):
    init_dist(rank, world_size, 29512)
    model = build_model_under_zero_init()
    simulate_bad_checkpoint_load(model, rank)
    for buf in model.buffers():
        dist.broadcast(buf, 0)
    out = run_forward_with_gather(model, get_shared_input())
    report_diffs("[FIX 1 - broadcast buffers]", rank, world_size, out)


def case_fix2_exclude_from_zero(rank, world_size, _port):
    init_dist(rank, world_size, 29513)
    torch.manual_seed(1234)
    model = VisionTokenizer().to(get_accelerator().current_device_name())  # NOT under zero.Init()
    with torch.no_grad():
        model.bn.running_mean.fill_(5.0)
        model.bn.running_var.fill_(2.0)
    model.eval()
    with torch.no_grad():
        out = model(get_shared_input())
    report_diffs("[FIX 2 - kept outside zero.Init]", rank, world_size, out)


def run(fn, world_size, master_port):
    mp.spawn(fn, args=(world_size, master_port), nprocs=world_size, join=True)


if __name__ == "__main__":
    run(case_bug, WORLD_SIZE, 29511)
    run(case_fix1_broadcast_buffers, WORLD_SIZE, 29512)
    run(case_fix2_exclude_from_zero, WORLD_SIZE, 29513)
```
